README.md
# The timber compiler

The Timber compiler implements Spar/Java, a superset of Java with
additional language constructs for high-performance computation. The
additions include support for data parallelism and task parallelism,
tuples, multi-dimensional arrays, generic classes and interfaces through
templates, complex numbers, and subscript operator overloading. See the
Timber website at https://cvr42.github.io/timber-compiler/index.html for
a detailed description of the compiler.

Timber is implemented as a static whole-program compiler, using a standard
C++ compiler to generate machine code. For now only Gnu C compilers are
supported, but that is mainly a restriction in our configuration system.
We plan to support other compilers in the future.

The Spar language extensions are only supported for source files with
the extension .spar.  For source files with the extension .java or .jav,
the Timber compiler behaves like a standard Java compiler. The main
restrictions are (a) no support for dynamic class loading (because we
do static whole-program compilation), and (b) no support for threads
(because we provide our own parallelization constructs).

Since the compiler uses well-defined and accessible internal data
structures, we believe that it is also very suitable as an experimental
compiler.

Timber is released under the Gnu Public License (GPL).

The Timber compiler has been designed to be portable, but it has mainly
been tested on recent Debian i386 Linux systems.  It requires two external
packages: a recent version of Tm, the template manager; and Kaffe version
1.0.6. Both can be downloaded from the page mentioned above.


compiler/ChangeLog
This file gives an overview of the changes between the releases of
the Timber compiler.

*** New features in Timber 2.0.1 compared to 2.0. ***

Timber 2.0.1 was released 13 december 2002

+ The elements of an array are no longer stored in a separately allocated
block, but starts immediately after the array descriptor.

+ When a program never catches a null pointer or bounds violation
exception, generate a special handler that does an exit() instead of
throwing an exception, and take this into account in in the program
analysis.

+ Other optimizations to generated code.

+ Improved language compatibility.

+ Added an 'uninstall' make rule.

+ Added support for MacOS X.



*** New features in Timber 2.0 compared to 1.6. ***

Timber 2.0 was released 28 march 2002

+ Official support for parallelization.

+ Further improvements for language compliance.


*** New features in Timber 1.6 compared to 1.5. ***

Timber 1.6 was released 21 November 2001.

+ Some modifications to the compiler to improve language compliance.

+ Some performance improvements.


*** New features in Timber 1.5 compared to 1.4. ***

Timber 1.5 was released 23 October 2001.

+ A lot of tuning has been done to improve the analyzer, resulting
in better performance for a larger set of programs. For exmaple,
an attempt is made to do static garbage collection, by letting the
compiler insert __delete statements for objects that are known to be
garbage. This currently almost never works for class objects, but for
arrays it works better. In many cases, significantly more efficient
programs are generated.

+ This version of the compiler uses the new analyzer treewalkers in
Tm version 2.0.8. It may be necessary to get a new version of Tm from
<http://www.pds.its.tudelft.nl/~reeuwijk/software/Tm/downloading.html>.

+ The Jacks compiler validation suite has been applied, and a lot
of small details have been modified to be more conformant with Java.
The current score is comparable to that of the Sun Java compiler.

+ The compiler now contains parallelization engines (in the directory
'engines'). The portability and documentation of these engines does
not yet meet our quality requirements, so for now these engines must
be considered experimental. We plan to release a working parallelizing
compiler in the near future.



website/news.pl
$webpage = $0;
$webpage =~ s/\.pl//;
require 'style.pli';
$title = "News about the Timber compiler";
&generateHeader( $webpage, $title );

$version_history_url = &generate_relative_url( $webpage, "spar/newfeatures.html" );
$download_url = &generate_relative_url( $webpage, "downloading.html" );
$jacks_url = &generate_relative_url( $webpage, "jacks.html" );
$requirements_url = &generate_relative_url( $webpage, "requirements.html" );
print <<EOF;
<h2>News items</h2>

EOF

&generate_news_item( $webpage, "13 December 2002", "Version 2.0.1 of the Timber compiler is available for <a href=\"$download_url\">downloading</a>. See the <a href=\"$version_history_url\">version history</a> page for a list of new features." ); 

&generate_news_item( $webpage, "28 March 2002", "Version 2.0 of the Timber compiler is available for <a href=\"$download_url\">downloading</a>. See the <a href=\"$version_history_url\">version history</a> page for a list of new features." ); 

&generate_news_item( $webpage, "21 November 2001", "Version 1.6 of the Timber compiler is available for <a href=\"$download_url\">downloading</a>. See the <a href=\"$version_history_url\">version history</a> page for a list of new features." ); 

&generate_news_item( $webpage, "23 October 2001", "Version 1.5 of the Timber compiler is available for <a href=\"$download_url\">downloading</a>. See the <a href=\"$version_history_url\">version history</a> page for a list of new features." ); 

&generate_news_item( $webpage, "23 October 2001", "A <a href=\"$jacks_url\">separate web page</a> now compares the language conformance of the Timber compiler with that of traditional compilers. The results may surprise you!" );

&generate_news_item( $webpage, "22 March 2001", "Version 1.4 of the Timber compiler is available for <a href=\"$download_url\">downloading</a>. See the <a href=\"$version_history_url\">version history</a> page for a list of new features. Please note that this release requires Kaffe version 1.0.6. This version is <a href=\"$requirements_url\">also available</a> for downloading." );

&generate_news_item( $webpage, "14 Feb 2001", "The website has been overhauled completedly. New features include a more obvious structure of the website, and this news page." );

&generate_news_item( $webpage, "24 Jan 2001", "Version 1.3 of the Timber compiler is available for <a href=\"$download_url\">downloading</a>. See the <a href=\"$version_history_url\">version history</a> page for a list of new features." );

print <<EOF;
<h2>Timber announcements mailing list</h2>
<p>
If you want to keep informed about the development of the Timber 
compiler, you may want to subscribe to the announcements mailing list.
This is a low-volume mailing list to announce updates of the Timber
compiler.
</p>
<p>
You can subscribe or unsubscribe to the mailing list by
clicking one of the buttons below:
</p>
 
<form action="mailto:$timberAnnounceRequestEmailAddress" method="post" enctype="text/plain">
<p> 
<input type="submit" name="subscribe " value=" Subscribe" />
<input type="submit" name="unsubscribe " value=" Unsubscribe" />
</p>
</form>
<p>
Note: this does not work on all web browsers, and may require the
installation of additional software.
</p>
<p> 
Alternatively, send an email with the subject <tt>subscribe</tt>
or <tt>unsubscribe</tt> to the email address <a
href="mailto:$timberAnnounceRequestEmailAddress?subject=subscribe">$timberAnnounceRequestEmailAddress</a>.
</p>

EOF

&generateFooter($0);
